File: apps/web/src/utils/libraryQuery.js
/** Shared vocabulary for the server-driven Library: sort menu, facet labels,
 * view definitions, and the translation from URL state to v2 API parameters.
 *
 * Every value here must have a server counterpart in
 * apps/api/backend/services/library_query.py — the page never filters locally.
 */
import { healthFacet } from './documentTags';
import {
    editionOf,
    familyKeyFromName,
    familyTitleFromKey,
    sortEditions,
} from './editions';

import { isoDateDaysAgo } from './time';

export const DEFAULT_SORT = 'name';

/** Grouped sort menu. `relevance` is offered only while a search is active. */
export const SORT_GROUPS = [
    {
        label: null,
        options: [{ value: 'relevance', label: 'Most relevant', searchOnly: true }],
    },
    {
        label: 'Name',
        options: [
            { value: 'name', label: 'Title A → Z' },
            { value: 'name_desc', label: 'Title Z → A' },
        ],
    },
    {
        label: 'Dates',
        options: [
            { value: 'newest', label: 'Recently added' },
            { value: 'oldest', label: 'Oldest added' },
            { value: 'updated', label: 'Recently updated' },
            { value: 'year', label: 'Edition — newest first' },
            { value: 'year_asc', label: 'Edition — oldest first' },
        ],
    },
    {
        label: 'Size',
        options: [
            { value: 'pages', label: 'Pages — most first' },
            { value: 'pages_asc', label: 'Pages — fewest first' },
            { value: 'sections', label: 'Sections — most first' },
            { value: 'sections_asc', label: 'Sections — fewest first' },
        ],
    },
    {
        label: 'Review',
        options: [
            { value: 'completion', label: 'Review progress' },
            { value: 'flagged', label: 'Most flagged' },
            { value: 'health', label: 'Health — outside gate first' },
        ],
    },
];

export const SORT_OPTIONS = SORT_GROUPS.flatMap((group) => group.options);
export const SORT_VALUES = new Set(SORT_OPTIONS.map((option) => option.value));

/** Name sorts group editions alphabetically; every other sort keeps server order. */
export function isNameSort(sort) {
    return sort === 'name' || sort === 'name_desc';
}

export function sortLabel(sort, { searching = false } = {}) {
    if (sort === 'relevance' && !searching) return sortLabel(DEFAULT_SORT);
    return SORT_OPTIONS.find((option) => option.value === sort)?.label || 'Title A → Z';
}

export const VIEWS = [
    { value: 'all', label: 'All documents' },
    { value: 'favorites', label: 'Favorites' },
    { value: 'recent', label: 'Recently viewed' },
];
export const VIEW_VALUES = new Set(VIEWS.map((view) => view.value));

export const KIND_LABELS = {
    'native-digital': 'Native digital',
    'scanned-ocr': 'Scanned (OCR)',
    'mixed-ocr': 'Mixed OCR',
    unknown: 'Unknown',
};

export const HEALTH_LABELS = {
    within_gate: 'Within gate',
    outside_gate: 'Outside gate',
    unmeasured: 'Unmeasured',
};

export const REVIEW_LABELS = {
    complete: 'Complete',
    in_progress: 'In progress',
    untouched: 'Untouched',
};

export const TAG_LABELS = {
    'native-digital': 'Native digital',
    'scanned-ocr': 'Scanned (OCR)',
    'mixed-ocr': 'Mixed OCR',
    'ocr-provisional': 'OCR provisional',
    'ocr-needs-review': 'OCR needs review',
    'pdf-inferred': 'Inferred from PDF',
};

export const ADDED_PRESETS = [
    { value: '7d', label: 'Last 7 days', days: 7 },
    { value: '30d', label: 'Last 30 days', days: 30 },
    { value: '90d', label: 'Last 90 days', days: 90 },
];

export const PAGE_PRESETS = [
    { value: 'small', label: 'Under 50 pages', min: null, max: 49 },
    { value: 'medium', label: '50 – 200 pages', min: 50, max: 200 },
    { value: 'large', label: '200 – 1000 pages', min: 200, max: 1000 },
    { value: 'huge', label: 'Over 1000 pages', min: 1001, max: null },
];

export function pagePresetFor(min, max) {
    return PAGE_PRESETS.find(
        (preset) => preset.min === (min ?? null) && preset.max === (max ?? null),
    )?.value || '';
}

export function tagLabel(tag) {
    return TAG_LABELS[tag] || tag;
}

/**
 * Translate library state to v2 API params. Returns `{ page, facets }` URLSearchParams
 * pair; `page` is null when a client-side view (favorites/recents) is empty — the
 * caller must render an empty result without a request. The facets params carry
 * `ids=-` in that case so counts collapse to zero while library-wide totals still
 * come back for the header strip.
 */
export function buildApiParams(state, { favoriteIds = [], recentIds = [] } = {}) {
    const { query, sort, view, facets } = state;
    const params = new URLSearchParams();
    const q = String(query || '').trim();
    if (q) params.set('q', q);
    if (facets.lanes.length) params.set('lane', facets.lanes.join(','));
    if (facets.kinds.length) params.set('kind', facets.kinds.join(','));
    if (facets.health.length) params.set('health', facets.health.join(','));
    if (facets.review.length) params.set('review', facets.review.join(','));
    if (facets.flagged) params.set('flagged', '1');
    if (facets.annotations) params.set('annotations', '1');
    if (facets.years.length) params.set('year', facets.years.join(','));
    if (facets.yearFrom != null) params.set('year_from', String(facets.yearFrom));
    if (facets.yearTo != null) params.set('year_to', String(facets.yearTo));
    const addedAfter = facets.addedPreset
        ? isoDateDaysAgo(ADDED_PRESETS.find((preset) => preset.value === facets.addedPreset)?.days || 0)
        : facets.addedAfter;
    if (addedAfter) params.set('added_after', addedAfter);
    if (facets.addedBefore) params.set('added_before', facets.addedBefore);
    if (facets.pagesMin != null) params.set('pages_min', String(facets.pagesMin));
    if (facets.pagesMax != null) params.set('pages_max', String(facets.pagesMax));
    if (facets.tags.length) params.set('tag', facets.tags.join(','));

    const effectiveSort = sort === 'relevance' && !q ? DEFAULT_SORT : sort;
    if (effectiveSort && effectiveSort !== DEFAULT_SORT) params.set('sort', effectiveSort);

    let pageParams = params;
    if (view === 'favorites' || view === 'recent') {
        const ids = (view === 'favorites' ? favoriteIds : recentIds).slice(0, 500);
        const facetsParams = new URLSearchParams(params);
        facetsParams.set('ids', ids.length ? ids.join(',') : '-');
        if (!ids.length) return { page: null, facets: facetsParams };
        pageParams = new URLSearchParams(params);
        pageParams.set('ids', ids.join(','));
        return { page: pageParams, facets: facetsParams };
    }
    return { page: pageParams, facets: params };
}

/** Number of active filter dimensions — the badge on the Filters button. */
export function countActiveFilters(facets) {
    return (
        facets.lanes.length
        + facets.kinds.length
        + facets.health.length
        + facets.review.length
        + (facets.flagged ? 1 : 0)
        + (facets.annotations ? 1 : 0)
        + (facets.years.length || facets.yearFrom != null || facets.yearTo != null ? 1 : 0)
        + (facets.addedPreset || facets.addedAfter || facets.addedBefore ? 1 : 0)
        + (facets.pagesMin != null || facets.pagesMax != null ? 1 : 0)
        + facets.tags.length
    );
}

export function docCompletion(doc) {
    const total = doc.total_sections || 0;
    if (total <= 0) return 0;
    return Math.round(((doc.stats?.reviewed || 0) / total) * 100);
}

/**
 * Group filtered documents by statute family.
 * ``documents`` should already be sorted by ``filterDocuments``; that order is
 * preserved across and within groups for non-name sorts. Name sorts keep the
 * friendlier year-within-family ordering (Z→A only reverses family titles).
 *
 * @returns {{ familyKey: string, title: string, editions: Array, outsideGate: boolean }[]}
 */
export function groupDocumentsByFamily(documents, sort = DEFAULT_SORT) {
    const orderIndex = new Map(documents.map((doc, index) => [doc.id, index]));
    const map = new Map();
    for (const doc of documents) {
        // The server's canonical family, not a second guess at it. Measured on the
        // real corpus, `familyKeyFromName` splits 5 of 29 server families into 34
        // groups -- the 21-edition Income Tax Ordinance among them -- because its
        // unanchored `dated` pattern eats the "UP" of "UPDATED UPTO". The backend
        // fixed and documented that; this copy never got it.
        const key = doc.family_key || familyKeyFromName(doc.name);
        if (!map.has(key)) {
            map.set(key, []);
        }
        map.get(key).push(doc);
    }
    const groups = [];
    const nameSort = isNameSort(sort);
    for (const [familyKey, docs] of map.entries()) {
        const editions = nameSort
            ? sortEditions(docs)
            : [...docs].sort(
                (a, b) => (orderIndex.get(a.id) ?? 0) - (orderIndex.get(b.id) ?? 0),
            );
        const outsideGate = editions.some((doc) => healthFacet(doc.health) === 'outside_gate');
        groups.push({
            familyKey,
            title: docs[0]?.family_title || familyTitleFromKey(familyKey),
            editions,
            outsideGate,
            latestYear: editionOf(
                nameSort ? editions[editions.length - 1] : editions[0],
            ).label,
            _order: Math.min(...docs.map((doc) => orderIndex.get(doc.id) ?? 0)),
        });
    }

    const sorted = nameSort
        ? groups.sort((a, b) => {
            const cmp = a.title.localeCompare(b.title);
            return sort === 'name_desc' ? -cmp : cmp;
        })
        : groups.sort((a, b) => a._order - b._order);

    return sorted.map(({ _order, ...group }) => group);
}

File: apps/api/backend/services/library_query.py
"""Shared filter/sort construction for the v2 Library endpoints.

The Library page is fully server-driven: every facet, sort, and the search box map
onto the SQL built here, so a paginated cursor can never disagree with the rows a
filter would have produced. The CASE expressions deliberately mirror the client
derivations in ``apps/web/src/utils`` (``documentLane``, ``healthFacet``,
``reviewFacet``, ``editionDateFromName``) — change one side, change the other.

All queries are flat (LATERAL aggregates instead of GROUP BY) so that filters,
sorting, COUNT(*), and LIMIT/OFFSET pagination compose without a grouping pass.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from backend.services.corpus_lanes import LANE_ORDER

SOURCE_KINDS = ("native-digital", "scanned-ocr", "mixed-ocr")
HEALTH_FACETS = ("within_gate", "outside_gate", "unmeasured")
REVIEW_FACETS = ("complete", "in_progress", "untouched")
DOC_STATUSES = ("pending", "in_progress", "blocked", "approved")
MAX_IDS = 500

# Resolved browse lane, mirroring corpus_lanes.classify_lane for rows whose stored
# corpus_lane is missing/unknown: title heuristics apply to acts_corpus rows only,
# everything else is a manual upload. Keep in sync with corpusLanes.js.
_TITLE = "regexp_replace(lower(COALESCE(d.name, '')), '\\s+', ' ', 'g')"
_LANES = ", ".join(f"'{lane}'" for lane in LANE_ORDER)
LANE_SQL = f"""
CASE
    WHEN d.corpus_lane IN ({_LANES}) THEN d.corpus_lane
    WHEN d.source_type <> 'acts_corpus' THEN 'manual'
    WHEN {_TITLE} ~ '\\y(rules?|regulations?)\\y' THEN CASE
        WHEN {_TITLE} LIKE '%income tax rule%' THEN 'income_tax_rules'
        WHEN {_TITLE} LIKE '%sales tax rule%'
            OR {_TITLE} LIKE '%sales tax special procedure%' THEN 'sales_tax_rules'
        WHEN {_TITLE} LIKE '%customs rule%' THEN 'customs_rules'
        WHEN {_TITLE} LIKE '%federal excise rule%' THEN 'federal_excise_rules'
        ELSE 'other_rules'
    END
    WHEN {_TITLE} LIKE '%income tax ordinance%' THEN 'ordinance'
    WHEN {_TITLE} LIKE '%customs act%' THEN 'customs'
    WHEN {_TITLE} LIKE '%sales tax act%' THEN 'sales_tax'
    WHEN {_TITLE} LIKE '%federal excise%' THEN 'federal_excise'
    WHEN {_TITLE} LIKE '%tax laws%' AND {_TITLE} LIKE '%amendment%' THEN 'tax_laws_amendment'
    WHEN {_TITLE} LIKE '%finance supplementary%' OR {_TITLE} ~ '\\yfinance act\\y' THEN 'finance'
    ELSE 'other_acts'
END
"""

# OCR/source kind from the provenance JSON blob. IS JSON guards the cast so a
# malformed or missing blob degrades to NULL ("unknown") instead of failing.
KIND_SQL = "CASE WHEN d.provenance IS JSON THEN CAST(d.provenance AS jsonb) ->> 'source_kind' END"

# Mirrors healthFacet in documentTags.js.
HEALTH_SQL = """
CASE
    WHEN m.measured_at IS NULL THEN 'unmeasured'
    WHEN m.gate_ok IS TRUE THEN 'within_gate'
    WHEN m.gate_ok IS FALSE THEN 'outside_gate'
    WHEN m.invariants_total IS NOT NULL AND m.invariants_total > 0 THEN
        CASE WHEN m.invariants_passed = m.invariants_total THEN 'within_gate' ELSE 'outside_gate' END
    ELSE 'unmeasured'
END
"""

# Mirrors reviewFacet in documentTags.js.
REVIEW_SQL = """
CASE
    WHEN d.total_sections <= 0 OR stats.reviewed <= 0 THEN 'untouched'
    WHEN stats.reviewed >= d.total_sections THEN 'complete'
    ELSE 'in_progress'
END
"""

# Edition year: identity-confirmed date first, else the first 19xx/20xx token in
# the name (the common path of editionDateFromName). Never raises on odd input.
# substring() returns the FIRST capturing group, so the group must wrap the year.
YEAR_SQL = """
CAST(COALESCE(
    substring(d.edition_date FROM '^(\\d{4})'),
    substring(d.name FROM '((19|20)\\d{2})')
) AS INTEGER)
"""

# One row per document; review stats, version recency, and health arrive as plain
# columns so WHERE/ORDER BY can use them directly.
FROM_DOCUMENTS = """
FROM documents d
LEFT JOIN LATERAL (
    SELECT
        COUNT(*) FILTER (WHERE s.review_status <> 'pending') AS reviewed,
        COUNT(*) FILTER (WHERE s.review_status = 'approved') AS approved,
        COUNT(*) FILTER (WHERE s.review_status = 'has_issues') AS has_issues,
        COUNT(*) FILTER (WHERE s.review_status = 'pending') AS pending,
        (SELECT COUNT(*) FROM annotations a
         WHERE a.document_id = d.id AND a.status = 'open') AS open_annotations
    FROM sections s
    WHERE s.document_id = d.id
) stats ON TRUE
LEFT JOIN LATERAL (
    SELECT COUNT(*) AS version_count, MAX(dv.created_at) AS last_version_at
    FROM document_versions dv
    WHERE dv.document_id = d.id
) vv ON TRUE
LEFT JOIN document_versions v ON v.document_id = d.id AND v.is_active = TRUE
LEFT JOIN version_metrics m ON m.version_id = v.id
"""

# Column list matching what _document_response reads (routes/documents.py).
PAGE_SELECT = """
SELECT
    d.id, d.name, d.pdf_filename, d.json_filename, d.total_sections,
    d.total_pages, d.uploaded_at, d.status, d.source_type, d.source_key,
    d.provenance, d.corpus_lane, d.withdrawn_at,
    stats.reviewed, stats.approved, stats.has_issues, stats.pending,
    stats.open_annotations,
    vv.version_count, vv.last_version_at,
    v.version_no AS active_version_no,
    v.id AS active_version_id,
    m.invariants_passed, m.invariants_total, m.cases_passed, m.cases_total,
    m.body_conserved, m.body_missing, m.footnote_conserved, m.footnote_missing,
    m.gate_ok, m.measured_at, m.detail_json
"""


# Provenance tag array, guarded so non-JSON / non-array shapes degrade to empty.
TAGS_SQL = """
CASE WHEN d.provenance IS JSON
        AND jsonb_typeof(CAST(d.provenance AS jsonb) -> 'tags') = 'array'
     THEN CAST(d.provenance AS jsonb) -> 'tags'
     ELSE CAST('[]' AS jsonb) END
"""


@dataclass(frozen=True)
class LibraryFilters:
    """Normalized filter state shared by the page, count, and facets queries."""

    q: str = ""
    lanes: tuple = ()
    kinds: tuple = ()
    health: tuple = ()
    review: tuple = ()
    flagged: bool = False
    annotations: bool = False
    status: str = ""
    years: tuple = ()
    year_from: Optional[int] = None
    year_to: Optional[int] = None
    added_after: str = ""
    added_before: str = ""
    pages_min: Optional[int] = None
    pages_max: Optional[int] = None
    tags: tuple = ()
    ids: tuple = ()
    # Withdrawn documents are out of the Library by default: the pipeline no longer
    # produces them, so listing them beside current ones invites a reviewer to
    # approve a parse that has been retired. They are never deleted, so this is a
    # filter and not a fact about the row -- set it to see them.
    include_withdrawn: bool = False

    def fingerprint_parts(self) -> tuple:
        return (
            self.q, self.lanes, self.kinds, self.health, self.review,
            self.flagged, self.annotations, self.status,
            self.years, self.year_from, self.year_to,
            self.added_after, self.added_before,
            self.pages_min, self.pages_max, self.tags, self.ids,
            self.include_withdrawn,
        )


def _in_clause(values: tuple) -> str:
    return "(" + ", ".join("?" for _ in values) + ")"


def build_where(filters: LibraryFilters, exclude: Optional[str] = None):
    """WHERE clause + params. ``exclude`` drops one dimension so facet counts for
    that dimension still respect every *other* active filter."""
    clauses: list[str] = []
    params: list = []

    def skipped(name: str) -> bool:
        return exclude == name

    if filters.q and not skipped("q"):
        clauses.append("(d.name ILIKE ? OR d.pdf_filename ILIKE ?)")
        needle = f"%{filters.q}%"
        params.extend((needle, needle))
    if filters.lanes and not skipped("lane"):
        clauses.append(f"{LANE_SQL} IN {_in_clause(filters.lanes)}")
        params.extend(filters.lanes)
    if filters.kinds and not skipped("kind"):
        known = [kind for kind in filters.kinds if kind != "unknown"]
        parts = []
        if known:
            parts.append(f"{KIND_SQL} IN {_in_clause(tuple(known))}")
            params.extend(known)
        if "unknown" in filters.kinds:
            parts.append(f"{KIND_SQL} IS NULL")
        clauses.append("(" + " OR ".join(parts) + ")")
    if filters.health and not skipped("health"):
        clauses.append(f"{HEALTH_SQL} IN {_in_clause(filters.health)}")
        params.extend(filters.health)
    if filters.review and not skipped("review"):
        clauses.append(f"{REVIEW_SQL} IN {_in_clause(filters.review)}")
        params.extend(filters.review)
    if filters.flagged and not skipped("flags"):
        clauses.append("stats.has_issues > 0")
    if filters.annotations and not skipped("flags"):
        clauses.append("stats.open_annotations > 0")
    if filters.status and not skipped("status"):
        clauses.append("d.status = ?")
        params.append(filters.status)
    if not skipped("year"):
        if filters.years:
            clauses.append(f"{YEAR_SQL} IN {_in_clause(filters.years)}")
            params.extend(filters.years)
        if filters.year_from is not None:
            clauses.append(f"{YEAR_SQL} >= ?")
            params.append(filters.year_from)
        if filters.year_to is not None:
            clauses.append(f"{YEAR_SQL} <= ?")
            params.append(filters.year_to)
    if not skipped("dates"):
        if filters.added_after:
            clauses.append("LEFT(d.uploaded_at, 10) >= ?")
            params.append(filters.added_after)
        if filters.added_before:
            clauses.append("LEFT(d.uploaded_at, 10) <= ?")
            params.append(filters.added_before)
    if not skipped("pages"):
        if filters.pages_min is not None:
            clauses.append("d.total_pages >= ?")
            params.append(filters.pages_min)
        if filters.pages_max is not None:
            clauses.append("d.total_pages <= ?")
            params.append(filters.pages_max)
    if filters.tags and not skipped("tags"):
        # jsonb_exists is the function form of the ? operator — which would clash
        # with this codebase's qmark parameter style.
        clauses.append(
            "(" + " OR ".join(f"jsonb_exists({TAGS_SQL}, ?)" for _ in filters.tags) + ")"
        )
        params.extend(filters.tags)
    if filters.ids and not skipped("ids"):
        clauses.append(f"d.id IN {_in_clause(filters.ids)}")
        params.extend(filters.ids)
    # Deliberately NOT skippable by `exclude`: a facet count that includes withdrawn
    # documents would not match the page it labels.
    if not filters.include_withdrawn:
        clauses.append("d.withdrawn_at IS NULL")

    where = " WHERE " + " AND ".join(clauses) if clauses else ""
    return where, params


def where_needs_stat_joins(filters: LibraryFilters, exclude: Optional[str] = None) -> bool:
    """True when the WHERE clause reads stats / health / review (LATERAL joins)."""
    def skipped(name: str) -> bool:
        return exclude == name

    return bool(
        (filters.health and not skipped("health"))
        or (filters.review and not skipped("review"))
        or (filters.flagged and not skipped("flags"))
        or (filters.annotations and not skipped("flags"))
    )


def count_from(filters: LibraryFilters, exclude: Optional[str] = None) -> str:
    """FROM for COUNT(*). Unfiltered / name-only filters do not need per-document
    section scans; those LATERAL joins are what made Library time out on a
    single API worker against a real corpus."""
    return FROM_DOCUMENTS if where_needs_stat_joins(filters, exclude) else "FROM documents d"


SORTS = {
    "name": "lower(d.name), d.id",
    "name_desc": "lower(d.name) DESC, d.id",
    "newest": "d.uploaded_at DESC, d.id",
    "oldest": "d.uploaded_at, d.id",
    "updated": "vv.last_version_at DESC NULLS LAST, d.id",
    "year": f"{YEAR_SQL} DESC NULLS LAST, lower(d.name), d.id",
    "year_asc": f"{YEAR_SQL} ASC NULLS LAST, lower(d.name), d.id",
    "pages": "d.total_pages DESC, d.id",
    "pages_asc": "d.total_pages, d.id",
    "sections": "d.total_sections DESC, d.id",
    "sections_asc": "d.total_sections, d.id",
    "completion": (
        "CAST(stats.reviewed AS float) / NULLIF(d.total_sections, 0) DESC NULLS LAST, "
        "lower(d.name), d.id"
    ),
    "flagged": "stats.has_issues DESC, lower(d.name), d.id",
    "health": (
        f"CASE WHEN {HEALTH_SQL} = 'outside_gate' THEN 0 "
        f"WHEN {HEALTH_SQL} = 'within_gate' THEN 1 ELSE 2 END, lower(d.name), d.id"
    ),
    "risk": (
        "CASE d.status WHEN 'blocked' THEN 0 WHEN 'in_progress' THEN 1 ELSE 2 END, "
        "lower(d.name), d.id"
    ),
}

SORT_VALUES = tuple(SORTS) + ("relevance",)


def order_sql(sort: str, q: str = ""):
    """ORDER BY clause + extra bind params (relevance needs the query text)."""
    if sort == "relevance" and q:
        return (
            "CASE WHEN d.name ILIKE (? || '%') THEN 0 ELSE 1 END, "
            "NULLIF(POSITION(lower(?) IN lower(COALESCE(d.name, ''))), 0) NULLS LAST, "
            "lower(d.name), d.id",
            (q, q),
        )
    return SORTS.get(sort, SORTS["name"]), ()
